submariner/globalnet-enhancement2-0.md

```diff
@@ -166,6 +168,53 @@ type EgressIPStatus struct {
     // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
     GlobalIPs []string `json:"globalIPs"`
 }
+
+type GlobalIngressIP struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec is the specification of desired behavior of GlobalIngressIP object.
+    Spec GlobalIngressIPSpec `json:"spec"`
+
+    // Observed status of GlobalIngressIP. Its a read-only field.
+    // +optional
+    Status GlobalIngressIPStatus `json:"status,omitempty"`
+}
+
+type GlobalIngressIPSpec struct {
+    // TypeOfObject can be ClusterIPService or a Pod
+    TypeOfObject string `json:"typeOfObject"`
+
+    // If the TypeOfObject points to a Pod that belongs to a Headless Service, it uses the
+    // name of the corresponding Service.
+    // +Optional
+    ServiceName string `json:"serviceName,omitempty"`
+}
```

Comment:
How will the relationship between `GlobalIngressIP` and `Service`/`Pod` be stored?
If it relies on that namespace and name are the same between resources, I'm worrying about the conflict on name collision of `Service` and `Pod`. The same resource with the same namespace and name can't be created, while different resource with the same namespace and name can be created. As a result, if `Service` and `Pod` have the same namespace and name, creation of `GlobalIngressIP` will fail.

Simple example is:
```
kubectl create -f https://raw.githubusercontent.com/kubernetes/website/master/content/en/examples/application/web/web.yaml
kubectl run nginx --image=nginx
kubectl expose pod nginx --port=80 --name=web-0

kubectl get all
NAME        READY   STATUS    RESTARTS   AGE
pod/nginx   1/1     Running   0          33s
pod/web-0   1/1     Running   0          4m51s
pod/web-1   1/1     Running   0          4m47s

NAME                 TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
service/kubernetes   ClusterIP   10.0.0.1     <none>        443/TCP   3h25m
service/nginx        ClusterIP   None         <none>        80/TCP    4m52s
service/web-0        ClusterIP   10.0.0.129   <none>        80/TCP    4s

NAME                   READY   AGE
statefulset.apps/web   2/2     4m52s
```

Reply:
@mkimuram In such a scenario, why would you need IngressIP for that Pod? As you mentioned in earlier comment, analogous for IngressIP is Endpoints. In scenario you shared, how does k8s handle naming conflict for Endpoints? If the `pod/nginx` doesn't have an Endpoint, why would it need an IngressIP?

Reply:
It is a good question,

Headless services would need an GlobalIngresIP I believe @sridhargaddam can confirm

Could we append a postfix to auto-generated resources, for example web-0-svc GlobalIngressIP, or web-0-pod ? 

We could also apply a randomization pattern postfix to pod generated GlobalIngressIP only (as k8s does to deployment pods )

Reply:
That's a good question @mkimuram. 

Yes @mangelajo, currently the GlobalIngressIP object will be created for ClusterIP Service (only one object) and for every Backend Pod in a Headless Service backed by Deployment or StatefulSets.

@mkimuram, as you pointed out, the issue could be seen while Globalnet tries to create a GlobalIngressIP object if we just use the name of Service or Pod alone. I discussed with @vthapar offline and we felt that we can prepend `svc-` or `pod-` while naming the `GlobalIngressIP` objects.

i.e., for the example where we have a ClusterIPService named `web-0` and HeadlessService with two backendPods named `web-0` and `web-1`
```
kubectl get globalIngressIPs -n default
svc-web-0
pod-web-0
pod-web-1
```

Along with this, to avoid ambiguity, I'll update the constants (which is an internal field) as shown below. 
```
const (
    ClusterIPService         TargetType = "ClusterIPService"
    HeadlessServicePod   TargetType = "HeadlessServicePod"
)
```

Note: GlobalIngressIP.Spec.ServiceName will contain the name of the corresponding Service.

Reply:
Thank you for sharing the idea.

I agree that it will work in most cases. There will be one corner case where length of the name + prefix exceeds the [maximum length](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names). In this case, the name can still collide.

I guess that we can go ahead with the idea by just adding additional fallback logic (something like, truncate name for long one, add suffix, match with patterns, and check if it is really for the resource). For that purpose, I think that `v1.LocalObjectReference` should be added to `GlobalIngressIPSpec` (Just adding name as string will also work, for we already have `TargetType`. Also, `v1.TypedLocalObjectRef` or `v1.ObjectReference` may be too much for this purpose).

I think that this fallback implementation can be TODO with name length restriction until it will be used in production, but it would be better to add the field to API with this case in mind.

Reply:
Done, thanks @mkimuram